engine/src/flutter/shell/platform/darwin/macos/framework/Source/ResizeSynchronizer.swift

```diff
@@ -0,0 +1,151 @@
+///     the native resize event. While blocked, it continuously polls for engine messages to process
+///     relevant updates. The `notify` closure is invoked just before blocking begins.
+///
+/// 2.  Commit (Raster Thread): As the engine renders frames, the raster thread, upon presenting a
+///     new frame, calls `performCommit(forSize:afterDelay:notify:)`, providing the size of the
+///     frame just rendered.
+///
+/// 3.  Unblocking: If the size of the committed frame matches the target size that `beginResize()`
+///     is waiting for, `beginResize()` unblocks. This allows the native resize operation to
+///     complete, now that Flutter's content is synchronized with the new window dimensions.
+///
+/// Safeguards:
+///   - Timeout: `beginResize()` includes a timeout mechanism to prevent indefinite blocking if a
+///     matching frame isn't committed in a timely manner. An optional `onTimeout` closure can be
+///     provided to handle this event.
+///   - Shutdown: The synchronization can be cleanly interrupted by calling `shutDown()`, which will
+///     also unblock any pending `beginResize()` call. After shutdown, `beginResize()` will no
+///     longer block.
+///   - Thread Safety: The class manages its internal state in a thread-safe manner to coordinate
+///     actions between the platform thread and the raster thread.
+@objc(FlutterResizeSynchronizer)
+public final class ResizeSynchronizer: NSObject {
+  private static let invalidSize = CGSize(width: -1, height: -1)
+
+  // Synchronizes access to _isInResize_unsafe: isInResize is accessed from multiple threads and
+  // thus requires synchronized access to the underlying storage.
+  private let isInResizeLock = NSLock()
+  private var _isInResize_unsafe: Bool = false
+  private var isInResize: Bool {
+    get {
+      isInResizeLock.lock()
+      defer { isInResizeLock.unlock() }
+      return _isInResize_unsafe
+    }
+    set {
+      isInResizeLock.lock()
+      _isInResize_unsafe = newValue
+      isInResizeLock.unlock()
+    }
+  }
+
+  // True if the app is shutting down. Must be set on platform thread.
+  private var isShuttingDown: Bool = false
+
+  // True if at least one frame has been presented. Must be set on platform thread.
+  private var didReceiveFrame: Bool = false
+
+  // The updated view surface size. Must be set on platform thread.
+  private var contentSize: CGSize = ResizeSynchronizer.invalidSize
+
+  /// Begins window resize operation to the specified size.
+  ///
+  /// Blocks the thread until `performCommit(forSize:notify:delay:)` with the same size is called.
+  /// While the thread is blocked, Flutter messages are being pumped.
+  /// See `FlutterRunLoop.mainRunLoop.pollFlutterMessagesOnce()`.
+  @objc public func beginResize(
+    forSize size: CGSize,
+    notify: () -> Void,
+    onTimeout: (() -> Void)? = nil
+  ) {
+    if !didReceiveFrame || isShuttingDown {
```

Comment:
Do we have a way to do "assert this is the main thread" here for debug builds only?

Reply:
FlutterRunLoop does this, but the existing code in ResizeSynchronizer did not. I think it could be reasonable to prevent people from accidentally calling from the wrong thread, but probably better to do in a followup patch to reduce the number of logical changes in this patch, which is mostly just a mechanism for ensuring we have infra to build/test swift in the macOS embedder. 

Checking the raster thread might be trickier since in tests, we don't spin up the full set of named threads.

Reply:
Gotcha, and that sounds good to me!